plugins/hadoopsecuritymanager-yarn/src/azkaban/security/HadoopSecurityManager_H_2_0.java

```diff
@@ -547,44 +694,76 @@ public synchronized void prefetchToken(final File tokenFile,
 
-        logger.info(HiveConf.ConfVars.METASTORE_KERBEROS_PRINCIPAL.varname
-            + ": "
-            + hiveConf
-                .get(HiveConf.ConfVars.METASTORE_KERBEROS_PRINCIPAL.varname));
+          // start to process the user inputs.
+          for (String hcatLocation : extraHcatLocations) {
+            String normalizedHcatLocStr = hcatLocation.trim().toLowerCase();
+            logger.info("Pre-fetching metaStore token for hcat location: "
+                + normalizedHcatLocStr);
+            // exchange the alias for the thrift url, and make sure the process
+            // goes thru.
+            String thriftUrl = extraHcatMapping.get(normalizedHcatLocStr);
+            if (thriftUrl == null) {
+              // short cut if the alias user specified is not found.
+              throw new HadoopSecurityManagerException(
+                  "failed to lookup hcat location for '"
+                      + normalizedHcatLocStr
+                      + "', the alias is not found in the system supported mapping list."
+                      + "All the supported aliases are :"
+                      + extraHcatMapping.keySet().toString());
+            }
 
-        HiveMetaStoreClient hiveClient = new HiveMetaStoreClient(hiveConf);
-        String hcatTokenStr =
-            hiveClient.getDelegationToken(userToProxy, UserGroupInformation
-                .getLoginUser().getShortUserName());
-        Token<DelegationTokenIdentifier> hcatToken =
-            new Token<DelegationTokenIdentifier>();
-        hcatToken.decodeFromUrlString(hcatTokenStr);
-        logger.info("Created hive metastore token: " + hcatTokenStr);
-        logger.info("Token kind: " + hcatToken.getKind());
-        logger.info("Token id: " + hcatToken.getIdentifier());
-        logger.info("Token service: " + hcatToken.getService());
-        cred.addToken(hcatToken.getService(), hcatToken);
-      } catch (Exception e) {
-        logger.error(
-            "Failed to get hive metastore token." + e.getMessage()
-                + e.getCause(), e);
+            // we are in the game, go and fetch the token for user.
+            logger.info("Resolved thrift url for hcat location '"
+                + normalizedHcatLocStr + "':" + thriftUrl);
+
+            hiveConf = new HiveConf();
+            hiveConf.set(HiveConf.ConfVars.METASTOREURIS.varname, thriftUrl);
+            hcatToken =
+                fetchHcatToken(userToProxy, hiveConf, normalizedHcatLocStr,
+                    logger);
+            cred.addToken(hcatToken.getService(), hcatToken);
+          }
+
+        }
+
+      } catch (HadoopSecurityManagerException hsme) {
+        throw hsme;
       } catch (Throwable t) {
-        logger.error(
+        String message =
             "Failed to get hive metastore token." + t.getMessage()
-                + t.getCause(), t);
+                + t.getCause();
+        logger.error(message, t);
+        throw new HadoopSecurityManagerException(message);
```

Comment:
HadoopSecurityManagerException doesn't have a ctor. override that takes in exception.